srtcore/fec.h

```diff
@@ -0,0 +1,248 @@
+        // distance between these two sequences, divided by
+        // rowsize*colsize should return %index-in-column,
+        // /number-series. The latter multiplied by the row size
+        // is the offset between the firstmost column and the
+        // searched column.
+        std::deque<RcvGroup> colq;
+
+        // This keeps the value of "packet received or not".
+        // The sequence number of the first cell is rowq[0].base.
+        // When dropping a row,
+        // - the firstmost element of rowq is removed
+        // - the length of one row is removed from this std::vector
+        int32_t cell_base;
+        std::deque<bool> cells;
+
+        // Note this function will automatically extend the container
+        // with empty cells if the index exceeds the size, HOWEVER
+        // the caller must make sure that this index isn't any "crazy",
+        // that is, it fits somehow in reasonable ranges.
+        bool CellAt(size_t index)
+        {
+            if (index >= cells.size())
+            {
+                // Cells not prepared for this sequence yet,
+                // so extend in advance.
+                cells.resize(index+1, false);
+                return false; // It wasn't marked, anyway.
+            }
+
+            return cells[index];
+        }
+
+        typedef SrtPacket PrivPacket;
+        std::vector<PrivPacket>& rebuilt;
+    } rcv;
+
+    void ConfigureGroup(Group& g, int32_t seqno, size_t gstep, size_t drop);
+    template <class Container>
+    void ConfigureColumns(Container& which, int32_t isn);
+
+    void ResetGroup(Group& g);
+
+    // Universal
+    void ClipData(Group& g, uint16_t length_net, uint8_t kflg,
+            uint32_t timestamp_hw, const char* payload, size_t payload_size);
+    void ClipPacket(Group& g, const CPacket& pkt);
+
+    // Sending
+    bool CheckGroupClose(Group& g, size_t pos, size_t size);
+    void PackControl(const Group& g, signed char groupix, SrtPacket& pkt, int32_t seqno);
+
+    // Receiving
+    void CheckLargeDrop(int32_t seqno);
+    int ExtendRows(int rowx);
+    int ExtendColumns(int colgx);
+    void MarkCellReceived(int32_t seq);
+    bool HangHorizontal(const CPacket& pkt, bool fec_ctl, loss_seqs_t& irrecover);
+    bool HangVertical(const CPacket& pkt, signed char fec_colx, loss_seqs_t& irrecover);
+    void ClipControlPacket(Group& g, const CPacket& pkt);
+    void ClipRebuiltPacket(Group& g, Receive::PrivPacket& pkt);
+    void RcvRebuild(Group& g, int32_t seqno, Group::Type tp);
+    int32_t RcvGetLossSeqHoriz(Group& g);
+    int32_t RcvGetLossSeqVert(Group& g);
+
+    static void TranslateLossRecords(const std::set<int32_t> loss, loss_seqs_t& irrecover);
+    void RcvCheckDismissColumn(int32_t seqno, int colgx, loss_seqs_t& irrecover);
+    int RcvGetRowGroupIndex(int32_t seq);
+    int RcvGetColumnGroupIndex(int32_t seq);
+    void CollectIrrecoverRow(RcvGroup& g, loss_seqs_t& irrecover);
+    bool IsLost(int32_t seq);
```

Comment:
I believe some of these functions can be constant.

Reply:
Right. I marked two of them. Many can't be though - even if the name suggests that it's going to only obtain some data, they are doing inside silently lazy container extension.